fastparquet/api.py

```diff
@@ -497,17 +497,20 @@ def _dtypes(self, categories=None):
             if dt.kind in ['i', 'b']:
                 # int/bool columns that may have nulls become float columns
                 num_nulls = 0
-                for rg in self.row_groups:
-                    chunk = rg.columns[i]
-                    if chunk.meta_data.statistics is None:
-                        num_nulls = True
-                        break
-                    if chunk.meta_data.statistics.null_count is None:
-                        num_nulls = True
-                        break
-                    if chunk.meta_data.statistics.null_count:
-                        num_nulls = True
-                        break
+                if not self.schema.is_required(col):
+                    num_nulls = True
+                else:
```

Comment:
In fact, the num_nulls tests below will never run for now for any column that could have nulls, so in this version it is wasted code. 

The initial check could be `'.' in col and not self.schema.is_required(col)`, i.e., we only act so pessimistically in this special nested case, but keep the old behaviour for normal columns. This means that an optional int column will stay as int in the case that we can be certain it has no nulls.